staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go

```diff
@@ -209,12 +209,55 @@ func (r *responder) Error(_ http.ResponseWriter, _ *http.Request, err error) {
 
 // these methods provide locked access to fields
 
-func (r *proxyHandler) updateAPIService(apiService *apiregistrationapi.APIService) {
+// updateWhileKeepingAvailability updates an api service handling info
+// if this is the first time we see the service we take availability directly from the apiService argument
+// otherwise we take it from the currently stored value
+//
+// note that the current availability of the service is updated through updateAvailabilityIfExists method
+func (r *proxyHandler) updateWhileKeepingAvailability(apiService *apiregistrationapi.APIService) {
+	serviceAvailable := apiregistrationapi.IsAPIServiceConditionTrue(apiService, apiregistrationapi.Available)
+	value := r.handlingInfo.Load()
+	if value != nil {
+		handlingInfo := value.(proxyHandlingInfo)
+		serviceAvailable = handlingInfo.serviceAvailable
+	}
+
+	r.updateAPIService(apiService, serviceAvailable)
+}
+
+// updateAvailabilityIfExists updates availability of a service
+// this method is noop if availability of a service hasn't changed or no service has been registered
+func (r *proxyHandler) updateAvailabilityIfExists(apiService *apiregistrationapi.APIService) {
+	serviceAvailable := apiregistrationapi.IsAPIServiceConditionTrue(apiService, apiregistrationapi.Available)
+	value := r.handlingInfo.Load()
+	if value == nil {
+		return
+	}
+
+	handlingInfo := value.(proxyHandlingInfo)
+	if handlingInfo.serviceAvailable == serviceAvailable {
+		// if availability of the service hasn't changed this is noop
+		return
+	}
+	r.updateAPIService(apiService, serviceAvailable)
+}
+
+func (r *proxyHandler) updateAPIService(apiService *apiregistrationapi.APIService, serviceAvailable bool) {
 	if apiService.Spec.Service == nil {
 		r.handlingInfo.Store(proxyHandlingInfo{local: true})
 		return
 	}
 
+	newInfo, err := r.newProxyHandlingInfo(apiService)
+	if err != nil {
+		klog.Warning(newInfo.transportBuildingError.Error())
```

Comment:
what does an error mean here?

Reply:
why can we ignore it?

Reply:
The error is not ignored it will be reported to the callers, here https://github.com/kubernetes/kubernetes/blob/6e31183d4780ec340af59c333dca6163ac79ad18/staging/src/k8s.io/kube-aggregator/pkg/apiserver/handler_proxy.go#L119

I think this is how we can report broken service to the callers if we ignore the error, they might never know why the service was broken.